Tabview tabbar horizontal padding for contentContainer

Freddy03h opened the pull request:
Use `padding(Horizontal|Left|Right)` values on `contentContainerStyle` for better styling on TabBar in `react-native-tab-view`

**Motivation**

Useful to create a custom style or adding insets values on left and right padding on contentContainer.
Use case is similar to existing `gap` prop (that I also fixed in this PR)

It's also fix #8667 

**Test plan**

I use TabViews existing examples, I adding `gap` and `contentContainerStyle` `paddingHorizontal` to AutoWidthTabBar, CustomIndicator, ScrollableTabBar and TabBarIcon.

okwasniewski reviewed (changes requested):
Thanks for a huge amount of great work! 🎉

I've tested the new behaviour on every platform, however I've observed some issues with RTL in the auto width example. I'm attaching a recording: 

https://github.com/react-navigation/react-navigation/assets/52801365/1725b1d7-90b8-4c63-a7f8-069fabaf25aa

I've also left some comments to make the code a bit cleaner. Once the above issues are resolved we can merge your changes.

Freddy03h commented on example/src/Screens/TabView/AutoWidthTabBar.tsx:
```diff
@@ -46,8 +46,10 @@ export const AutoWidthTabBar = () => {
+      contentContainerStyle={styles.tabbarContentContainer}
       labelStyle={styles.label}
       tabStyle={styles.tabStyle}
+      gap={40}
```
Ok I'll fix RTL and change gap size examples

Freddy03h commented on example/src/Screens/TabView/CustomIndicator.tsx:
```diff
@@ -73,16 +84,19 @@ export const CustomIndicator = () => {
           {
-            width: `${100 / navigationState.routes.length}%`,
+            width: width,
             transform: [{ translateX }] as any,
```
It's not a line I changed, or relative to this PR, do you want me to change it anyway?

okwasniewski replied:
Right, let's leave it I will change it in another PR 👍🏻

Freddy03h commented on packages/react-native-tab-view/src/TabBar.tsx:
```diff
@@ -82,13 +83,50 @@ const getFlattenedTabWidth = (style: StyleProp<ViewStyle>) => {
   return tabStyle?.width;
 };
 
+const getFlattenedPaddingLeft = (style: StyleProp<ViewStyle>) => {
+  const flattenStyle = StyleSheet.flatten(style);
+
+  return flattenStyle
+    ? flattenStyle.paddingLeft || flattenStyle.paddingHorizontal || 0
+    : 0;
+};
+
+const getFlattenedPaddingRight = (style: StyleProp<ViewStyle>) => {
+  const flattenStyle = StyleSheet.flatten(style);
+
+  return flattenStyle
+    ? flattenStyle.paddingRight || flattenStyle.paddingHorizontal || 0
+    : 0;
+};
```
The point of `getFlattenedPadding(Left|Right)` is to retrieve the "real" padding and not only the style prop value.

So I'll prefer to keep those functions and use `getFlattenedStyleProperty` inside:
```js
const getFlattenedPaddingLeft = (style: StyleProp<ViewStyle>) => {
  return getFlattenedStyleProperty('paddingLeft', style) || getFlattenedStyleProperty('paddingHorizontal', style) || 0;
};
```

Freddy03h replied:
I'm thinking… there is a downside of `getFlattenedStyleProperty` : There will be more call of `StyleSheet.flatten` and it can tax in terms of optimizations https://reactnative.dev/docs/stylesheet#flatten

Freddy03h commented on packages/react-native-tab-view/src/TabBar.tsx:
```diff
@@ -82,13 +83,50 @@ const getFlattenedTabWidth = (style: StyleProp<ViewStyle>) => {
+    : 0;
+};
+
+const convertPaddingPercentToSize = (
+  value: FlattenedTabPadding,
+  layout: Layout
+): number => {
+  switch (typeof value) {
+    case 'number':
+      return value;
+    case 'string':
+      if (value.endsWith('%')) {
+        const width = parseFloat(value);
+        if (Number.isFinite(width)) {
+          return layout.width * (width / 100);
+        }
+      }
+  }
+  return 0;
+};
```
I matched the style of `getComputedTabWidth` to avoid refactoring untouched code and keep consistency (but otherwise, I agree with you with readability)

Freddy03h commented:
Hi, I fixed RTL on auto width, reduced gap and padding examples, fix types and commented on some code style / refactor messages. Is there anything else I can do?

okwasniewski commented:
Hey, 

Sorry for the delay. I've just tested the app and unfortunately found some new issues with this PR: 
- On iOS the TabBar layout shifts on mount in auto width example
- Coverflow example is broken

https://github.com/react-navigation/react-navigation/assets/52801365/b838b0b7-f6ab-4036-8bd1-2f4b7fb91a2e

Freddy03h commented:
Hi, thank's for the review.

I tried on main branch and RTL issues already existed, I fixed the one on auto width tabbar because I found it so… I fixed it, but it's kinda out of scope of this PR…
I don't plan to fixed RTL coverflow example because the issue is already on main.

I found the issue with layout shift on auto width, it was a tricky one because it's also the cause of an issue without gap setted 😅 Thank you to noticed it! I didn't saw it with the auto refresh behavior…

okwasniewski commented:
Thanks for addressing the comments 🙌

Freddy03h commented:
Thank you for the merge 👍
